app/main.py

```diff
@@ -1,34 +1,126 @@
+from dataclasses import dataclass, field
+
+
+@dataclass
 class Deck:
-    def __init__(self, row, column, is_alive=True):
-        pass
 
+    row: int
+    column: int
+    is_alive: bool = True
 
+
+@dataclass
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    start: tuple
+    end: tuple
+    is_drowned: bool = False
+    decks: list = field(init=False)
+
+    def __post_init__(self) -> None:
+        decks = []
+        for x_coord in range(self.start[0], self.end[0] + 1):
+            for y_coord in range(self.start[1], self.end[1] + 1):
+                decks.append(Deck(x_coord, y_coord))
+        self.decks = decks
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def get_deck(self, row: int, column: int) -> Deck:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, column: int) -> None:
+        deck = self.get_deck(row, column)
+        deck.is_alive = False
+
+        hit_decks = 0
+        for deck in self.decks:
+            if deck.is_alive is False:
+                hit_decks += 1
+        if hit_decks == len(self.decks):
+            self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list) -> None:
+        self._validate_field(ships)
+
+        field = {}
+        for x_coord in range(10):
+            for y_coord in range(10):
+                field[(x_coord, y_coord)] = None
+
+        for item in ships:
+            ship = Ship(item[0], item[1])
+            for deck in ship.decks:
+                field[(deck.row, deck.column)] = ship
+
+        self.field = field
+
+    @staticmethod
+    def _validate_field(ships: list) -> None:
+
+        if len(ships) != 10:
+            raise ValueError("There must be 10 ships.")
+
+        for ship in ships:
+            other_ships = [
+                end for temp in ships if temp != ship for end in temp
```

Comment:
The list comprehension here is incorrect. The variable `end` is used before assignment. It should be `for temp in ships if temp != ship for end in temp`.